### PR TITLE
Prototypes [WIP]

### PR DESCRIPTION
**Pull request by AnturK**
## Prototypes

I felt science was lacking a sense of discovery for some time, and while experimentor was a good try it was too niche to be used much.
### How do you make one
1. Get prototype chassis - Found in test lab or printed with high tier R&D
2. Get at least 3 protoboards - Found in maintenance, from disassembling strange devices, or ordered through cargo crate.
3. Slot 3 (For just the effect) or 6 protoboards into the chassis,you will be notified if the selected boards will create an effect.
### What do they do

Few Examples : 
AAC BDA Prototype - Rusts walls in a line in front of the prototype
EAC BDA Prototype - Lubes all floors in front of the prototype
CDB ABB Prototype - Makes all lizardmen on station drop their hats.

AAC = CAA = ACA The order of the boards is not important at the moment.
### Screenshots

space station 13 2015-07-27 124115
ui
### Todo:
- [x] Add at least 30 effects (~ 12 done)
- [x] Add few targeting types
- [x] Cargo crate/maint spawns for the protoboards
- [x] Logging
- [x] Sprites
- [ ] Pretty up the code - this is rough draft that i just made for some quick heh prototyping
### Effects right now:

Affected by targeting:
- Makes humans hungry
- Makes specific hallucinations
- Rusts walls
- Lubes floors
- Injects ethanol
- Injects space drugs
- Injects mannitol
- Paints turfs (blue,red,green)
- Makes humans scream
- Makes explosion sounds
- Tames simple animals
- Shoots fireworks
- Stops bleeding
- Takes hats off
- Makes hydroponic trays fully watered and fertilized
- Bolts doors
- Raises corpses as zombies
- Creates cigars
- Grows cat ears on humans
- Powers APCs
- Gives hugs
- Fixes borgs
- Fills beakers (milk)
- Mines
- Breaks windows
- Makes webs
- Grows grass

Not affected by targeting
- Creates gas clouds (full of carbon)
- Electrocutes you
- Pulls all metal items to it
- Creates flashbang level noise
- Augments chests
- Predicts future
- Detects Ghosts
- Creates portals to other dimensions (they spew enemies/ change turfs around them with time)

Targeting modes:
- Turfs directly around the device and the user (Default)
- Range (2,5,7)
- View (7)
- Line on the other side of the user (cannon)
- Line perpedicular to the user (wall)
- Circle around the machine
- All humans on the zlevel (no turfs)
- All lizards on the zlevel (no turfs)
- turfs in + shape around all mobs in view

### CODE REVIEW

**Comment by Rinoka** [comment]
EAC BDA Prototype - Lubes all floors in front of the prototype

image

**Comment by Rinoka** [comment]
but no seriously though i can approve of this, just avoid making this a thing where traitors can make a hellish slip 'n slide into space when the shuttle is called

**Comment by Iamgoofball** [comment]
oh shiiiit

:+1:

**Comment by Chiefwaffles** [comment]
There should be some way to basically turn a completed prototype into research via some method. Maybe by doing something like using it X number of times, or successfully describing its function; something that requires use of the prototype to complete. So you could make a prototype, find out how it works, then go back to your lab and increase a research level or two because of that.

**Comment by Incoming5643** [comment]
one of the biggest failings the experimentor had was that it only had a few practical effects, and the artifacts it made had only a few predictable types.

So you should either have either a TON of possible effects or more practical uses (or both)

**Comment by Kelenius** [comment]
I can see all the wizards getting wrecked by the drophat thing.

File: code/modules/prototypes/prototype.dm [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+/datum/prototype/effect/screamer/human_effect(var/mob/living/carbon/human/target)
+	target.emote("scream")
+
+/datum/prototype/effect/boombox/human_effect(var/mob/living/carbon/human/target)
+	target << 'sound/effects/Explosion1.ogg'
+
+/datum/prototype/effect/tamer/mob_effect(var/mob/target)
+	target.faction |= "neutral"
+	target << "<span class='notice'>You feel calm.</span>"
```
Comment (Dorsisdwarf) [inline code/modules/prototypes/prototype.dm]: Uhhhh
Won't this deconvert nuke ops and stuff?

Comment (Iamgoofball) [inline code/modules/prototypes/prototype.dm]: Only the npc ones.

**Comment by AnturK** [comment]
I'll try to write up something for the research return. It's not bad idea.

I'm trying to aim for 2:1 ratio between gimmicky and useful/harmful effects. There's also cooldown depending on effect (and i'll probably add range cooldown multiplier) so the more powerful effects can't be abused (too much)

**Comment by freerealestate** [comment]
seems kind of silly, pointless and overcomplicated imo

**Comment by Dorsisdwarf** [comment]
:+1: 
Just so long as this stays important and supported and isn't Experimentor 2: Shitastic Boogaloo

**Comment by Supermichael777** [comment]
so the first three are effect and the second group is targeting. certain combos may need to have a chance of silly/dangerous feedback (violently ejecting the boards for example)

**Comment by Aranclanos** [comment]
>commiting wips
see you in two months

**Comment by AnturK** [comment]
Currently 36/56 combinations works for effects 11/56 for targets, they share same combination space though.

File: code/modules/prototypes/prototype.dm [hunk 2]
```diff
@@ -0,0 +1,875 @@
...
+	turf_paths = list(/turf/simulated/floor/engine/cult)
+	mob_paths =  list(/mob/living/simple_animal/hostile/faithless) //replace with constructs when they get turned into /hostile
+	mobs_per_wave = 1
+/obj/effect/rift/hivebot
+	turf_paths = list(/turf/simulated/floor/bluegrid,/turf/simulated/floor/greengrid)
+	mob_paths =  list(/mob/living/simple_animal/hostile/hivebot,/mob/living/simple_animal/hostile/hivebot/range,/mob/living/simple_animal/hostile/hivebot/strong)
+	mobs_per_wave = 3
+
+/obj/effect/rift/alien
+	turf_paths = list(/turf/simulated/floor/plating/asteroid)
```
Comment (Anonus) [inline code/modules/prototypes/prototype.dm]: This would be better if it spawned resin floors.

Comment (AnturK) [inline code/modules/prototypes/prototype.dm]: Yeah, but there's no turf equivalent for these. I suppose i could just drop some nodes around

File: code/modules/prototypes/prototype.dm [hunk 3]
```diff
@@ -0,0 +1,875 @@
...
+			user << "<span class='notice'>The [owner] points west.</span>"
+			M.Turn(-90)
+			I.transform = M
+		if("E")
+			user << "<span class='notice'>The [owner] points east.</span>"
+			M.Turn(90)
+			I.transform = M
+	flick_overlay(I,list(user.client),50)
+	return
+
```
Comment (MrPerson) [inline code/modules/prototypes/prototype.dm]: Abso-fucking-lutely not

Comment (Iamgoofball) [inline code/modules/prototypes/prototype.dm]: ???

Comment (duncathan) [inline code/modules/prototypes/prototype.dm]: >empty return at the end of proc

**Comment by Razharas** [comment]
Theres literally nothing wrong with PRing one line of code and then finishing it all in the course of months
Trust me i did it many many many times

**Comment by AnturK** [comment]
I was off the net for a week, will be working/gutting/closing this soon

File: code/modules/prototypes/prototype.dm (hunk 3 above)
Comment (Iamgoofball) [inline code/modules/prototypes/prototype.dm]: this isn't allowed?

**Pull request closed without merging**